### PR TITLE
Implementation of concept merge action in Django admin interface VGNWB-46

### PR DESCRIPTION
**Pull request by yyk16293**

### CODE REVIEW

File: concepts/admin.py [hunk 1]
```diff
@@ -20,30 +23,49 @@ def merge_concepts(modeladmin, request, queryset):
     :param queryset:
     :return:
     """
-    
-    opts = modeladmin.model._meta
-    app_label = opts.app_label
+    #return render_to_response('admin/Sample.html')
     RESOLVED_STATUS = Concept.RESOLVED
```
Comment (erickpeirson) [inline concepts/admin.py]: Why define a new variable here?

File: concepts/admin.py [hunk 2]
```diff
@@ -20,30 +23,49 @@ def merge_concepts(modeladmin, request, queryset):
...
-        resolvedConceptName = resolvedConcept['label']
 
-        context = {
-        "resolvedConcept": resolvedConceptName,
-        "opts": opts,
-        "app_label": app_label,
-        "unResolvedConcepts": unResolvedConceptsList
-        }
-        return render_to_response('admin/merge_concepts_resolved.html', context)
+        for i in range(0,len(unResolvedConceptsList)):
```
Comment (erickpeirson) [inline concepts/admin.py]: `len()` will trigger a database call -- use the `.count()` method

File: concepts/admin.py [hunk 3]
```diff
@@ -20,30 +23,49 @@ def merge_concepts(modeladmin, request, queryset):
...
-        resolvedConceptsList = queryset.filter(concept_state = RESOLVED_STATUS).values()
-        unResolvedConceptsList = queryset.exclude(concept_state = RESOLVED_STATUS).values()
 
-        #As there will be only one element in the list
+    if 'post' in request.POST:
+            # process the queryset here
+
+        unResolvedConceptsList = queryset.exclude(concept_state = RESOLVED_STATUS).values()
+        resolvedConceptsList = queryset.filter(concept_state = RESOLVED_STATUS).values()
         resolvedConcept = resolvedConceptsList[0]
```
Comment (erickpeirson) [inline concepts/admin.py]: What if there are no resolved concepts?

File: concepts/admin.py [hunk 4]
```diff
@@ -20,30 +23,49 @@ def merge_concepts(modeladmin, request, queryset):
...
+            presentConcept = unResolvedConceptsList.get(i)
+            presentConcept.merged_with = resolvedConcept
+            presentConcept.concept_state = Concept.REJECTED
+
+        return render_to_response('admin/Sample.html')
+    else:
+        opts = modeladmin.model._meta
+        app_label = opts.app_label
+
+        resolvedConceptCount = queryset.filter(concept_state = RESOLVED_STATUS).count()
```
Comment (erickpeirson) [inline concepts/admin.py]: Why define a new variable `resolvedConceptCount` and a new queryset here? If you define the queryset `resolvedConceptsList` above the `if` block, you can just use  `resolvedConceptsList.count()` later on.

File: concepts/admin.py [hunk 5]
```diff
@@ -20,30 +23,49 @@ def merge_concepts(modeladmin, request, queryset):
...
+
+    #VGNWB-121 gets called only when there is only one resolved concept
+        if resolvedConceptCount == 1:
+            resolvedConceptsList = queryset.filter(concept_state = RESOLVED_STATUS).values()
```
Comment (erickpeirson) [inline concepts/admin.py]: Again, why define this again?

File: concepts/admin.py [hunk 6]
```diff
@@ -20,30 +23,49 @@ def merge_concepts(modeladmin, request, queryset):
...
     RESOLVED_STATUS = Concept.RESOLVED
-    resolvedConceptCount = queryset.filter(concept_state = RESOLVED_STATUS).count()
 
-    #VGNWB-121 gets called only when there is only one resolved concept
-    if resolvedConceptCount == 1:
-        resolvedConceptsList = queryset.filter(concept_state = RESOLVED_STATUS).values()
-        unResolvedConceptsList = queryset.exclude(concept_state = RESOLVED_STATUS).values()
 
-        #As there will be only one element in the list
+    if 'post' in request.POST:
```
Comment (erickpeirson) [inline concepts/admin.py]: Not sure why `'post'` would be in `request.POST`; are you trying to check whether the request is a POST request? If so, see [this snippet](https://docs.djangoproject.com/en/1.9/ref/request-response/#django.http.HttpRequest.method) for an example.

**Comment by erickpeirson** [comment]
@yyk16293 Good start, but you can really clean this code up quite a bit. See comments for a few places where things could be greatly simplified. The principle here is **D**on't **R**epeat **Y**ourself.

Also, can you give your PRs more informative names? "making required changes to integrate action behaviour" is pretty nebulous....

Thanks! :-)

File: concepts/admin.py [hunk 7]
```diff
@@ -20,30 +23,40 @@ def merge_concepts(modeladmin, request, queryset):
...
     """
-    
-    opts = modeladmin.model._meta
-    app_label = opts.app_label
-    RESOLVED_STATUS = Concept.RESOLVED
-    resolvedConceptCount = queryset.filter(concept_state = RESOLVED_STATUS).count()
+    unResolvedConceptsList = queryset.exclude(concept_state = Concept.RESOLVED).values()
+    resolvedConceptsList = queryset.filter(concept_state = Concept.RESOLVED).values()
+
+    if 'POST' in request.POST:
```
Comment (erickpeirson) [inline concepts/admin.py]: This still needs to be fixed; 

``` python
if request.method == 'POST':
```

File: concepts/admin.py [hunk 8]
```diff
@@ -11,47 +15,53 @@ def resolve(modeladmin, request, queryset):
         authorities.resolve(type(obj), obj)
 resolve.verbose_name = 'resolve selected concepts'
 
+class SeriesForm(forms.Form):
```
Comment (erickpeirson) [inline concepts/admin.py]: Two blank lines above top-level class and function declarations

File: concepts/admin.py [hunk 9]
```diff
@@ -11,47 +15,53 @@ def resolve(modeladmin, request, queryset):
...
+        _selected_action = forms.CharField(widget=forms.MultipleHiddenInput)
+
```
Comment (erickpeirson) [inline concepts/admin.py:20]: Two blank lines above top-level class and function declarations

File: concepts/admin.py [hunk 10]
```diff
@@ -11,47 +15,53 @@ def resolve(modeladmin, request, queryset):
...
-    RESOLVED_STATUS = Concept.RESOLVED
-    resolvedConceptCount = queryset.filter(concept_state = RESOLVED_STATUS).count()
-
-    #VGNWB-121 gets called only when there is only one resolved concept
-    if resolvedConceptCount == 1:
-        resolvedConceptsList = queryset.filter(concept_state = RESOLVED_STATUS).values()
-        unResolvedConceptsList = queryset.exclude(concept_state = RESOLVED_STATUS).values()
-
-        #As there will be only one element in the list
+
```
Comment (erickpeirson) [inline concepts/admin.py:25]: Please add a docstring that describes what this does

File: concepts/admin.py [hunk 11]
```diff
@@ -11,47 +15,53 @@ def resolve(modeladmin, request, queryset):
...
+            # process the queryset here
+            unResolvedConceptsList.update(concept_state=Concept.REJECTED)
+            unResolvedConceptsList.update(merged_with=resolvedConcept)
+
+        else:
+            opts = modeladmin.model._meta
+            app_label = opts.app_label
+            resolvedConceptCount = queryset.filter(concept_state = Concept.RESOLVED).count()
+
+            #VGNWB-121 gets called only when there is only one resolved concept
```
Comment (erickpeirson) [inline concepts/admin.py]: What's VGNWB-121? Someone reading the code won't know what that means -- refer to the function that you're talking about

File: concepts/admin.py [hunk 12]
```diff
@@ -11,47 +15,53 @@ def resolve(modeladmin, request, queryset):
...
-    return render_to_response('admin/merge_concepts.html', {'concepts': queryset})
+            return render_to_response('admin/merge_concepts.html', {'concepts': queryset})
 
 
 class ConceptAdmin(admin.ModelAdmin):
     model = Concept
     search_fields = ('label',)
     list_display = ('label', 'resolved', 'concept_state', 'typed',)
-    actions = (resolve, merge_concepts,)
+    actions = (resolve,merge_concepts)
```
Comment (erickpeirson) [inline concepts/admin.py]: space after comma

**Comment by erickpeirson** [comment]
@yyk16293 A few stylistic fixes; see line comments. Deploying this locally to test, so if you want to work on those I'll get back to you in a few minutes

File: concepts/admin.py [hunk 13]
```diff
@@ -11,47 +15,53 @@ def resolve(modeladmin, request, queryset):
...
-    #VGNWB-121 gets called only when there is only one resolved concept
-    if resolvedConceptCount == 1:
-        resolvedConceptsList = queryset.filter(concept_state = RESOLVED_STATUS).values()
-        unResolvedConceptsList = queryset.exclude(concept_state = RESOLVED_STATUS).values()
-
-        #As there will be only one element in the list
+
+        unResolvedConceptsList = queryset.exclude(concept_state = Concept.RESOLVED)
+        resolvedConceptsList = queryset.filter(concept_state = Concept.RESOLVED)
         resolvedConcept = resolvedConceptsList[0]
```
Comment (erickpeirson) [inline concepts/admin.py]: When I select two concepts that are not resolved, and select the "merge concepts" action, I get an IndexError here.

image

Comment (yyk16293) [inline concepts/admin.py]: Yes Erick, I figured out the mistake I am doing in the code. Working on that.

**Pull request merged**